substrate/client/rpc-spec-v2/src/archive/archive_storage.rs

```diff
@@ -0,0 +1,155 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Implementation of the `archive_storage` method.
+
+use std::{marker::PhantomData, sync::Arc};
+
+use sc_client_api::{Backend, ChildInfo, StorageKey, StorageProvider};
+use sp_api::BlockT;
+
+use crate::common::{
+	events::{
+		ArchiveStorageMethodErr, ArchiveStorageMethodOk, ArchiveStorageResult,
+		PaginatedStorageQuery, StorageQueryType,
+	},
+	storage::{is_key_queryable, IterQueryType, QueryIter, Storage},
+};
+
+/// Generates the events of the `chainHead_storage` method.
+pub struct ArchiveStorage<Client, Block, BE> {
+	/// Storage client.
+	client: Storage<Client, Block, BE>,
+	/// The maximum number of reported items by the `archive_storage` at a time.
+	storage_max_reported_items: usize,
+	/// The maximum number of queried items allowed for the `archive_storage` at a time.
+	storage_max_queried_items: usize,
+	_phantom: PhantomData<(BE, Block)>,
+}
+
+impl<Client, Block, BE> ArchiveStorage<Client, Block, BE> {
+	/// Constructs a new [`ArchiveStorage`].
+	pub fn new(
+		client: Arc<Client>,
+		storage_max_reported_items: usize,
+		storage_max_queried_items: usize,
+	) -> Self {
+		Self {
+			client: Storage::new(client),
+			storage_max_reported_items,
+			storage_max_queried_items,
+			_phantom: PhantomData,
+		}
+	}
+}
+
+impl<Client, Block, BE> ArchiveStorage<Client, Block, BE>
+where
+	Block: BlockT + 'static,
+	BE: Backend<Block> + 'static,
+	Client: StorageProvider<Block, BE> + 'static,
+{
+	/// Generate the response of the `archive_storage` method.
+	pub fn handle_query(
+		&self,
+		hash: Block::Hash,
+		mut items: Vec<PaginatedStorageQuery<StorageKey>>,
+		child_key: Option<ChildInfo>,
+	) -> ArchiveStorageResult {
+		if let Some(child_key) = child_key.as_ref() {
+			if !is_key_queryable(child_key.storage_key()) {
+				return ArchiveStorageResult::Ok(ArchiveStorageMethodOk {
+					result: Vec::new(),
+					discarded_items: 0,
+				})
+			}
+		}
+
+		let discarded_items = items.len().saturating_sub(self.storage_max_queried_items);
```

Comment:
Is `self.storage_max_queried_items` part of the spec? What is the value we use in production for this?

Reply:
As far as I can understand the spec, there's no explicit mention of an upper limit before the pagination kicks in. This is by design, since smoldot / other implementation might be running under hardware constraints.

For `chainHead_storage` we are using 5 items, but we haven't collect yet any data to make a better informed decision: 

https://github.com/paritytech/polkadot-sdk/blob/f4f1b23f055d186db8e6c80a7271ca5be0efdd39/substrate/client/rpc-spec-v2/src/chain_head/chain_head.rs#L85-L87

The `archive` class is not exposed by substrate yet, but would keep both classes in sync

Reply:
It would be to good to add a prometheus metric for that or some annoying log for us to know whether this limit is exceeded often or not ^^